File: 2015/ruby/day13/part2.rb
#!/usr/bin/ruby

def sit(sat, happiness)
    max = 0

    for i in 0..8
        next if sat.include?(i)
        max = [sit(sat.push(i), happiness), max].max
        sat.pop()
    end

    if sat.size == 9
        max += happiness[sat[0]][sat[8]] + happiness[sat[8]][sat[0]]
        for i in 1..8
            max += happiness[sat[i]][sat[i - 1]] + happiness[sat[i - 1]][sat[i]]
        end
    end

    return max
end

begin
    File.open('input.txt', 'r') do |file|
        people = {
            'Alice' => 0,
            'Bob' => 1,
            'Carol' => 2,
            'David' => 3,
            'Eric' => 4,
            'Frank' => 5,
            'George' => 6,
            'Mallory' => 7
        }
        happiness = Array.new(9) { |value| Array.new(9) { |value| 0 }}

        file.each_line do |line|
            sentence = line[0..-3].split(' ')
            happiness[people[sentence[0]]][people[sentence[10]]] = sentence[3].to_i

            if sentence[2] == 'lose'
                happiness[people[sentence[0]]][people[sentence[10]]] *= -1
            end
        end

        puts sit([], happiness)
    end
rescue SystemCallError
    puts 'input.txt not found.'
end

File: 2016/python/day22/part2.py
#!/usr/bin/python

def construct(line):
    line = line.translate(None, 'Txy').split()
    _, x, y = line[0].split('-')
    return {'pos': (int(x), int(y)), 'size': int(line[1]), 'used': line[2]}


with open('input.txt') as file_handler:
    nodes = [construct(line) for line in file_handler]

empty = next(node for node in nodes if node['used'] == '0')
wall = next(node for node in nodes if node['size'] > 500)

print empty['pos'][1] + 2 * (empty['pos'][0] - wall['pos'][0] + 1) + 176
